docs/configuration/squest_settings.md
# Configuration  settings

Default settings are configured to provide a testing/development environment. For a production setup it is recommended
to adjust them following your target environment.

When using docker-compose, the configuration is loaded from environment variables file placed in the folder `docker/environment_variables`.

When using Kubernetes, the configuration need to be placed in the `squest.yml` inventory file in the variable `squest_django/env`.

## Database

### DB_ROOT_PASSWORD

Set the database root/superuser password

### DB_DATABASE

**Default:** `squest_db`

Database name.

### DB_USER

**Default:** `squest_user`

User used to connect to the `DB_DATABASE` name.

### DB_PASSWORD

**Default:** `squest_password`

Password of the `DB_USER` username.  

### DB_HOST

**Default:** `127.0.0.1`

Database host. The default value is localhost to match the development configuration. 
Switch to `db` in production when using the docker-compose based deployment.

### DB_PORT

**Default:** `5432`

Default postgres database port.


## Authentication

### LDAP

#### LDAP_ENABLED

**Default:** `False`

Set to `True` to enable LDAP based authentication.  


#### AUTH_LDAP_SERVER_URI

**Default:** `ldap:port`

Set the LDAP serveur URI and port

#### AUTH_LDAP_BIND_DN

**Default:** `cn=service_account_name,ou=Applications,o=domain.com`

Set the LDAP DN to authenticate to the LDAP serveur

#### AUTH_LDAP_BIND_PASSWORD

**Default:** `NONE`

Associated to AUTH_LDAP_BIND_DN, password used to authenticate DN

#### AUTH_LDAP_USER_SEARCH

**Default:** `ou=People,o=domain.com`

User search patern

#### AUTH_LDAP_ATTR_FIRSTNAME

**Default:** `givenName`

set the LDAP "first_name" attribute

#### AUTH_LDAP_ATTR_LASTNAME

**Default:** `sn`

set the LDAP "last_name" attribute

#### AUTH_LDAP_ATTR_MAIL

**Default:** `mail`

set the LDAP "email" attribute


### OpenID Connect

#### SOCIAL_AUTH_OIDC_ENABLED

**Default:** `False`

set to `True` to enable OpenID Connect based authentication

#### SOCIAL_AUTH_OIDC_BTN_TEXT

**Default:** `OpenID Login`

set the OpenID Connect login button text

#### SOCIAL_AUTH_OIDC_OIDC_ENDPOINT

**Default:** `None`

set the endpoint to the base url and the remaining configuration will be auto-completed like so: `<SOCIAL_AUTH_OIDC_OIDC_ENDPOINT>/.well-known/openid-configuration`

#### SOCIAL_AUTH_OIDC_KEY

**Default:** `None`

set to the OpenID client ID that Squest should be configured to use

#### SOCIAL_AUTH_OIDC_SECRET

**Default:** `None`

set to the OpenID client secret that Squest should be configured to use

### Password

#### PASSWORD_ENABLED

**Default:** `True`

set to `False` to disable the password form on the login view

### DEFAULT_ADMIN_TOKEN

**Default:** `None`

Set an API token that will be linked to the admin user when starting Squest.

### DEFAULT_ADMIN_PASSWORD

**Default:** `None`

Set a default password to the admin user when starting Squest.

### LOGIN_HELPER_TEXT

**Default:** `None`

Add a custom note into the login page that helps user to know what are the expected credentials. HTML text is supported.

E.G: "Use your corporate email and password".

## Squest

### MAINTENANCE_MODE_ENABLED

**Default:** `False`

When enabled, only administrators can access squest UI and API. 
This can be used for example to block new requests by end users from the service catalog. So an administrator can perform operations against the API like migrating instance specs.

!!! note

    This can also be set on the fly from the admin panel (top right corner of the UI) in the object `Squest settings`.

### SQUEST_HOST

**Default:** `http://squest.domain.local`

Address of the Squest portal instance. Used in email templates and in metadata sent to Red Hat Ansible Automation Platform/AWX job templates.

### SQUEST_EMAIL_HOST

**Default:** `squest@squest.domain.local`

Domain name used as email sender. E.g: "squest@squest.domain.local". 

### SQUEST_ADMINS

**Default:** `''`

A list of all the email who get code error notifications. When DEBUG=False.
Example:

```text
elias.boulharts@mail.com,nicolas.marcq@mail.com
```

### SQUEST_EMAIL_NOTIFICATION_ENABLED

**Default:** Based on `DEBUG` value by default

Set to `True` to enable email notifications.  

### IS_DEV_SERVER

**Default:** `False`

Set to `True` to change the navbar and footer color to visually identify a testing instance of Squest.

### GUNICORN_WORKERS

**Default:** `4`

Number of workers used by Gunicorn process in charge of serving client connection. Increase the number of worker threads to serve more clients concurrently

## SMTP

### EMAIL_HOST

**Default:** `localhost`

The SMTP host to use for sending email.

### EMAIL_PORT

**Default:** `25`

Port to use for the SMTP server defined in `EMAIL_HOST`.

### EMAIL_HOST_USER

**Default:** `None`

User to use to authenticate with the SMTP server defined in `EMAIL_HOST` in combination with `EMAIL_HOST_PASSWORD`. Leave empty/unconfigured to send emails unauthenticated.

### EMAIL_HOST_PASSWORD

**Default:** `None`

Password to use to authenticate with the SMTP server defined in `EMAIL_HOST` in combination with `EMAIL_HOST_USER`. Leave empty/unconfigured to send emails unauthenticated.

### EMAIL_USE_SSL

**Default:** `False`

Whether to use an implicit TLS (secure) connection when talking to the SMTP server defined in `EMAIL_HOST`.

## Backup

### BACKUP_ENABLED

**Default:** `False`

Switch to `True` to enable backup. Refer to the [dedicated documentation](../administration/backup.md).

### BACKUP_CRONTAB

**Default:** `0 1 * * *`

Crontab line for backup. By default, the backup is performed every day at 1AM.

### DBBACKUP_CLEANUP_KEEP

**Default:** `5`

Number of db backup file to keep. [Doc](https://django-dbbackup.readthedocs.io/en/master/configuration.html#dbbackup-cleanup-keep-and-dbbackup-cleanup-keep-media).

### DBBACKUP_CLEANUP_KEEP_MEDIA

**Default:** `5`

Number of media backup tar to keep. [Doc](https://django-dbbackup.readthedocs.io/en/master/configuration.html#dbbackup-cleanup-keep-and-dbbackup-cleanup-keep-media).

## Metrics

### METRICS_ENABLED

**Default:** `False`

Switch to `True` to enable Prometheus metrics page.

### METRICS_PASSWORD_PROTECTED

**Default:** `True`

Switch to `False` to disable the basic authentication on metrics page.

### METRICS_AUTHORIZATION_USERNAME

**Default:** `admin`

Username for the basic authentication of the metrics page.

### METRICS_AUTHORIZATION_PASSWORD

**Default:** `admin`

Password for the basic authentication of the metrics page.

## Auto cleanup

### DOC_IMAGES_CLEANUP_ENABLED

**Default:** `False`

Switch to `True` to enable automatic cleanup of ghost docs images from media folder.

### DOC_IMAGES_CLEANUP_CRONTAB

**Default:** `30 1 * * *`

Crontab line for ghost image cleanup. By default performed every day at 1:30 AM.

## Production

### SECRET_KEY

**Default:**  Default randomly-generated

Django secret key used for cryptographic signing. [Doc](https://docs.djangoproject.com/en/4.2/ref/settings/#std:setting-SECRET_KEY).

### DEBUG

**Default:** `True`

Django DEBUG mode. Switch to `False` for production.

### ALLOWED_HOSTS

**Default:** `*`

Comma separated list of allowed FQDN. Refer to the complete [documentation](https://docs.djangoproject.com/en/4.2/ref/settings/#allowed-hosts).   

### CSRF_TRUSTED_ORIGINS

**Default:** `http://127.0.0.1:8080,http://localhost:8080`

Comma separated list of trusted origin. 
Refer to the complete [documentation](https://docs.djangoproject.com/en/4.2/ref/settings/#csrf-trusted-origins).

!!! note

    If you deploy with `http://1.2.3.4:8080`, please add `CSRF_TRUSTED_ORIGINS=http://127.0.0.1:8080,http://localhost:8080,http://1.2.3.4:8080` in your variables.

### CELERY_BROKER_URL

**Default:** `amqp://rabbitmq:rabbitmq@localhost:5672/squest`

RabbitMQ message broker URL. The default value is localhost to match the development configuration. 
Replace `localhost` by `rabbitmq` in production when using the docker-compose based deployment.

### LANGUAGE_CODE

**Default:** `en-us`

Django language. [Doc](https://docs.djangoproject.com/en/4.2/ref/settings/#language-code)

### TIME_ZONE

**Default:** `Europe/Paris`

Time zone of the server that host Squest service. [Doc](https://docs.djangoproject.com/en/4.2/ref/settings/#std:setting-TIME_ZONE)

### DATE_FORMAT

**Default:** `%d %b, %Y  %H:%M`

Change the format of all date in Squest UI. Based on Python [strftime](https://strftime.org/).

## Plugins

### FIELD_VALIDATOR_PATH

!!!warning

    FIELD_VALIDATOR_PATH is now deprecated. Please use [SURVEY_VALIDATOR_PATH](#survey_validator_path) instead.

**Default:** `plugins/field_validators`

Path to form field validation modules.

### SURVEY_VALIDATOR_PATH

**Default:** `plugins/survey_validators`

Path to SurveyValidator modules.

## Redis

### REDIS_CACHE_USER

**Default:** `default`

Username of Redis account.

### REDIS_CACHE_PASSWORD

**Default:** `redis_secret_password`

Password of Redis account.

### REDIS_CACHE_HOST

**Default:** `127.0.0.1`

Redis hostname.

### REDIS_CACHE_PORT

**Default:** `6379`

Redis port.


Squest/settings.py
"""
Django settings for Squest project.

Generated by 'django-admin startproject' using Django 3.1.7.

For more information on this file, see
https://docs.djangoproject.com/en/3.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.1/ref/settings/
"""
import os
import sys
from pathlib import Path
import git
from Squest.version import __version__

from celery.schedules import crontab

from service_catalog.utils import str_to_bool, get_mysql_dump_major_version, \
    get_celery_crontab_parameters_from_crontab_line

# LOAD configuration from the environment
SECRET_KEY = os.environ.get('SECRET_KEY', 'sxuxahnezvrrea2vp97et=q(3xmg6nk4on92+-+#_s!ikurbh-')
DEBUG = str_to_bool(os.getenv('DEBUG', True))
ALLOWED_HOSTS = os.getenv('ALLOWED_HOSTS', '*').split(',')
CSRF_TRUSTED_ORIGINS = os.getenv('CSRF_TRUSTED_ORIGINS', 'http://127.0.0.1:8080,http://localhost:8080').split(',')
DB_DATABASE = os.environ.get('DB_DATABASE', 'squest_db')
DB_USER = os.environ.get('DB_USER', 'squest_user')
DB_PASSWORD = os.environ.get('DB_PASSWORD', 'squest_password')
DB_HOST = os.environ.get('DB_HOST', '127.0.0.1')
DB_PORT = os.environ.get('DB_PORT', '5432')
LDAP_ENABLED = str_to_bool(os.environ.get('LDAP_ENABLED', False))
SOCIAL_AUTH_OIDC_ENABLED = str_to_bool(os.environ.get('SOCIAL_AUTH_OIDC_ENABLED', False))
SOCIAL_AUTH_OIDC_BTN_TEXT = os.environ.get('SOCIAL_AUTH_OIDC_BTN_TEXT', 'OpenID Login')
SOCIAL_AUTH_OIDC_OIDC_ENDPOINT = os.environ.get('SOCIAL_AUTH_OIDC_OIDC_ENDPOINT', '')
SOCIAL_AUTH_OIDC_KEY = os.environ.get('SOCIAL_AUTH_OIDC_KEY', '')
SOCIAL_AUTH_OIDC_SECRET = os.environ.get('SOCIAL_AUTH_OIDC_SECRET', '')
PASSWORD_ENABLED = str_to_bool(os.environ.get('PASSWORD_ENABLED', True))
BACKUP_ENABLED = str_to_bool(os.environ.get('BACKUP_ENABLED', False))
BACKUP_CRONTAB = os.environ.get('BACKUP_CRONTAB', "0 1 * * *")  # every day at 1 AM
DOC_IMAGES_CLEANUP_ENABLED = str_to_bool(os.environ.get('DOC_IMAGES_CLEANUP_ENABLED', False))
DOC_IMAGES_CLEANUP_CRONTAB = os.environ.get('DOC_IMAGES_CLEANUP', "30 1 * * *")  # every day at 1:30 AM.
REDIS_CACHE_USER = os.environ.get('REDIS_CACHE_USER', 'default')
REDIS_CACHE_PASSWORD = os.environ.get('REDIS_CACHE_PASSWORD', 'redis_secret_password')
REDIS_CACHE_HOST = os.environ.get('REDIS_CACHE_HOST', '127.0.0.1')
REDIS_CACHE_PORT = os.environ.get('REDIS_CACHE_PORT', '6379')
DEFAULT_ADMIN_TOKEN = os.environ.get('DEFAULT_ADMIN_TOKEN', None)
DEFAULT_ADMIN_PASSWORD = os.environ.get('DEFAULT_ADMIN_PASSWORD', None)
MAINTENANCE_MODE_ENABLED = str_to_bool(os.environ.get('MAINTENANCE_MODE_ENABLED', False))
DATE_FORMAT = os.environ.get('DATE_FORMAT', "%d %b, %Y  %H:%M")
LOGIN_HELPER_TEXT = os.environ.get('LOGIN_HELPER_TEXT', None)
IS_DEV_SERVER = str_to_bool(os.environ.get('IS_DEV_SERVER', False))
SQL_DEBUG = str_to_bool(os.environ.get('SQL_DEBUG', False))
# -------------------------------
# SQUEST CONFIG
# -------------------------------
BASE_DIR = Path(__file__).resolve().parent.parent
TESTING = sys.argv[1:2] == ['test']
IS_COLLECTING_STATIC = sys.argv[1:2] == ['collectstatic']
IS_MAKING_MIGRATION = sys.argv[1:2] == ['makemigrations']
IS_MIGRATING = sys.argv[1:2] == ['migrate']
IS_INSERTING_DEFAULT_DATA = sys.argv[1:2] == ['insert_default_data']
IS_GUNICORN_EXECUTION = False
if "gunicorn" in sys.argv[0]:
    IS_GUNICORN_EXECUTION = True

print(f"BASE_DIR: {BASE_DIR}")
print(f"DEBUG: {DEBUG}")
print(f"TESTING: {TESTING}")
print(f"IS_COLLECTING_STATIC: {IS_COLLECTING_STATIC}")
print(f"IS_MAKING_MIGRATION: {IS_MAKING_MIGRATION}")
print(f"IS_MIGRATING: {IS_MIGRATING}")
print(f"IS_INSERTING_DEFAULT_DATA: {IS_INSERTING_DEFAULT_DATA}")
print(f"ALLOWED_HOSTS: {ALLOWED_HOSTS}")
print(f"CSRF_TRUSTED_ORIGINS: {CSRF_TRUSTED_ORIGINS}")
print(f"IS_GUNICORN_EXECUTION: {IS_GUNICORN_EXECUTION}")
print(f"LDAP_ENABLED: {LDAP_ENABLED}")
print(f"SOCIAL_AUTH_OIDC_ENABLED: {SOCIAL_AUTH_OIDC_ENABLED}")
print(f"BACKUP_ENABLED: {BACKUP_ENABLED}")
print(f"SQL_DEBUG: {SQL_DEBUG}")

if BACKUP_ENABLED:
    print(f"BACKUP_CRONTAB: {BACKUP_CRONTAB}")

if DEBUG:
    print(os.environ)

# Application definition
INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'rest_framework',
    'django_celery_results',
    'django_celery_beat',
    'tempus_dominus',
    'django_node_assets',
    'django_filters',
    'drf_yasg',
    'taggit',
    'martor',
    'django_tables2',
    'dbbackup',
    'service_catalog',
    'resource_tracker_v2',
    'profiles',
    'monitoring',
    'cachalot',
    'django_cleanup.apps.CleanupConfig',  # should stay last to override delete method of our model
]

if SOCIAL_AUTH_OIDC_ENABLED and not TESTING:
    INSTALLED_APPS += [
        'social_django'
    ]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'Squest.middleware.maintenance.MaintenanceMiddleware'
]

ROOT_URLCONF = 'Squest.urls'
DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
TEMPLATES_DIR = str(BASE_DIR) + os.sep + 'templates'
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [TEMPLATES_DIR],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

if SOCIAL_AUTH_OIDC_ENABLED and not TESTING:
    TEMPLATES[0]['OPTIONS']['context_processors'].append('social_django.context_processors.backends')
    TEMPLATES[0]['OPTIONS']['context_processors'].append('social_django.context_processors.login_redirect')

WSGI_APPLICATION = 'Squest.wsgi.application'

# Database
# https://docs.djangoproject.com/en/3.1/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': "django.db.backends.postgresql",
        'NAME': DB_DATABASE,
        'USER': DB_USER,
        'PASSWORD': DB_PASSWORD,
        'HOST': DB_HOST,
        'PORT': DB_PORT,
    }
}

# Password validation
# https://docs.djangoproject.com/en/3.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# -----------------------------------------
# LDAP
# -----------------------------------------
if LDAP_ENABLED:
    from Squest.ldap_config import *

# -----------------------------------------
# Authentication
# -----------------------------------------
# do not use LDAP auth when testing
AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
    'Squest.utils.squest_rbac.SquestRBACBackend',
]
if LDAP_ENABLED and not TESTING:
    AUTHENTICATION_BACKENDS = (
        'django.contrib.auth.backends.ModelBackend',
        'django_auth_ldap.backend.LDAPBackend',
        'Squest.utils.squest_rbac.SquestRBACBackend',
    )
if SOCIAL_AUTH_OIDC_ENABLED and not TESTING:
    AUTHENTICATION_BACKENDS = (
        'django.contrib.auth.backends.ModelBackend',
        'social_core.backends.open_id_connect.OpenIdConnectAuth',
        'Squest.utils.squest_rbac.SquestRBACBackend',
    )

# Internationalization
# https://docs.djangoproject.com/en/3.1/topics/i18n/

LANGUAGE_CODE = os.environ.get('LANGUAGE_CODE', 'en-us')
TIME_ZONE = os.environ.get('TIME_ZONE', 'Europe/Paris')
USE_I18N = True
USE_L10N = True
USE_TZ = True

LOGIN_REDIRECT_URL = 'home'
LOGOUT_REDIRECT_URL = 'home'

# -----------------------------------------
# Static files https://docs.djangoproject.com/en/3.1/howto/static-files/
# -----------------------------------------

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'static')
STATICFILES_DIRS = [
    os.path.join(BASE_DIR, 'project-static')  # project statics
]
MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
MEDIA_URL = '/media/'

# -----------------------------------------
# NODE CONFIG https://github.com/whitespy/django-node-assets
# -----------------------------------------
STATICFILES_FINDERS = [
    'django_node_assets.finders.NodeModulesFinder',
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
]
NODE_PACKAGE_JSON = os.path.join(BASE_DIR, 'package.json')
NODE_MODULES_ROOT = os.path.join(BASE_DIR, 'node_modules')

# -----------------------------------------
# LOGGING CONFIG
# -----------------------------------------
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
            # 'format': '{pathname}:{lineno} {levelname} {asctime} {module} {message}',
            'format': '{levelname} {asctime} {module} {message}',
            'style': '{',
        },
        'simple': {
            'format': '{levelname} {message}',
            'style': '{',
        },
    },
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
            'formatter': 'verbose',
        },
        "mail_admins": {
            "level": "ERROR",
            "class": "django.utils.log.AdminEmailHandler",
            "include_html": True,
        },
    },
    'loggers': {
        '': {
            'handlers': ['console'],
            'level': "DEBUG",
        },
        'graphviz': {
            'handlers': ['console'],
            'level': "INFO",
        }
    },
}

# -----------------------------------------
# CELERY CONFIG
# -----------------------------------------
CELERY_BROKER_URL = os.environ.get('CELERY_BROKER_URL', 'amqp://rabbitmq:rabbitmq@localhost:5672/squest')
print(f"CELERY_BROKER_URL: {CELERY_BROKER_URL}")
# Add a five-minutes timeout to all Celery tasks.
CELERY_TASK_ALWAYS_EAGER = TESTING
CELERY_WORKER_HIJACK_ROOT_LOGGER = False
CELERY_ACCEPT_CONTENT = ['json', 'msgpack']
CELERY_TASK_SERIALIZER = 'json'
CELERY_RESULT_SERIALIZER = 'json'
CELERY_RESULT_BACKEND = 'django-db'
CELERY_BEAT_SCHEDULER = 'service_catalog.celery_beat_scheduler.DatabaseSchedulerWithCleanup'
CELERY_BEAT_SCHEDULE = {}

# -----------------------------------------
# Squest email config
# -----------------------------------------
ADMINS = [("", element) for element in os.environ.get('SQUEST_ADMINS', '').split(',') if element]
SQUEST_HOST = os.environ.get('SQUEST_HOST', "http://squest.domain.local")
SQUEST_EMAIL_HOST = os.environ.get('SQUEST_EMAIL_HOST', "squest@squest.domain.local")
SERVER_EMAIL = SQUEST_EMAIL_HOST
SQUEST_EMAIL_NOTIFICATION_ENABLED = str_to_bool(os.environ.get('SQUEST_EMAIL_NOTIFICATION_ENABLED', False))
EMAIL_HOST = os.environ.get('EMAIL_HOST', 'localhost')
EMAIL_PORT = os.environ.get('EMAIL_PORT', 25)
EMAIL_HOST_USER= os.environ.get('EMAIL_HOST_USER', None)
EMAIL_HOST_PASSWORD= os.environ.get('EMAIL_HOST_PASSWORD', None)
EMAIL_USE_SSL= os.environ.get('EMAIL_USE_SSL', False)

print(f"ADMINS: {ADMINS}")
print(f"SQUEST_HOST: {SQUEST_HOST}")
print(f"SQUEST_EMAIL_HOST: {SQUEST_EMAIL_HOST}")
print(f"SQUEST_EMAIL_NOTIFICATION_ENABLED: {SQUEST_EMAIL_NOTIFICATION_ENABLED}")
print(f"EMAIL_HOST: {EMAIL_HOST}")
print(f"EMAIL_PORT: {EMAIL_PORT}")
print(f"EMAIL_HOST_USER: {EMAIL_HOST_USER}")
print(f"EMAIL_USE_SSL: {EMAIL_USE_SSL}")

# -----------------------------------------
# Martor CONFIG https://github.com/agusmakmun/django-markdown-editor
# -----------------------------------------
# Choices are: "semantic", "bootstrap"
MARTOR_THEME = 'bootstrap'

# Global martor settings
# Input: string boolean, `true/false`
MARTOR_ENABLE_CONFIGS = {
    'emoji': 'true',  # to enable/disable emoji icons.
    'imgur': 'true',  # to enable/disable imgur/custom uploader.
    'mention': 'false',  # to enable/disable mention
    'jquery': 'true',  # to include/revoke jquery (require for admin default django)
    'living': 'false',  # to enable/disable live updates in preview
    'spellcheck': 'false',  # to enable/disable spellcheck in form textareas
    'hljs': 'true',  # to enable/disable hljs highlighting in preview
}

# To show the toolbar buttons
MARTOR_TOOLBAR_BUTTONS = [
    'bold', 'italic', 'horizontal', 'heading', 'pre-code',
    'blockquote', 'unordered-list', 'ordered-list',
    'link', 'image-link', 'image-upload', 'emoji',
    'direct-mention', 'toggle-maximize', 'help'
]

# To setup the martor editor with title label or not (default is False)
MARTOR_ENABLE_LABEL = False

# Markdownify
MARTOR_MARKDOWNIFY_FUNCTION = 'martor.utils.markdownify'  # default
MARTOR_MARKDOWNIFY_URL = '/martor/markdownify/'  # default

# Markdown extensions (default)
MARTOR_MARKDOWN_EXTENSIONS = [
    'markdown.extensions.extra',
    'markdown.extensions.nl2br',
    'markdown.extensions.smarty',
    'markdown.extensions.fenced_code',

    # Custom markdown extensions.
    'martor.extensions.urlize',
    'martor.extensions.del_ins',  # ~~strikethrough~~ and ++underscores++
    'martor.extensions.mention',  # to parse markdown mention
    'martor.extensions.emoji',  # to parse markdown emoji
    'martor.extensions.mdx_video',  # to parse embed/iframe video
    'martor.extensions.escape_html',  # to handle the XSS vulnerabilities
]

# Markdown Extensions Configs
MARTOR_MARKDOWN_EXTENSION_CONFIGS = {}

# Markdown urls
MARTOR_UPLOAD_PATH = 'doc_images/uploads'
MARTOR_UPLOAD_URL = '/api/uploader/'  # change to local uploader

# Maximum Upload Image
# 2.5MB - 2621440
# 5MB - 5242880
# 10MB - 10485760
# 20MB - 20971520
# 50MB - 5242880
# 100MB 104857600
# 250MB - 214958080
# 500MB - 429916160
MAX_IMAGE_UPLOAD_SIZE = 5242880  # 5MB

# Markdown Extensions
# MARTOR_MARKDOWN_BASE_EMOJI_URL = 'https://www.webfx.com/tools/emoji-cheat-sheet/graphics/emojis/'     # from webfx
MARTOR_MARKDOWN_BASE_EMOJI_URL = 'https://github.githubassets.com/images/icons/emoji/'  # default from github
MARTOR_MARKDOWN_BASE_MENTION_URL = 'https://python.web.id/author/'  # please change this to your domain

# If you need to use your own themed "bootstrap" or "semantic ui" dependency
# replace the values with the file in your static files dir
# MARTOR_ALTERNATIVE_JS_FILE_THEME = "semantic-themed/semantic.min.js"   # default None
# MARTOR_ALTERNATIVE_CSS_FILE_THEME = "semantic-themed/semantic.min.css" # default None
MARTOR_ALTERNATIVE_JQUERY_JS_FILE = "jquery/dist/jquery.min.js"  # default None

if DOC_IMAGES_CLEANUP_ENABLED:
    CELERY_BEAT_SCHEDULE["cleanup_martor"] = {
        "task": "service_catalog.tasks.task_cleanup_ghost_docs_images",
        "schedule": crontab(**get_celery_crontab_parameters_from_crontab_line(DOC_IMAGES_CLEANUP_CRONTAB)),
    }

# -----------------------------------------
# DJANGO REST FRAMEWORK CONFIG
# -----------------------------------------
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.SessionAuthentication',
        'Squest.api.authentication.TokenAuthentication',
    ),
    'DEFAULT_FILTER_BACKENDS': ['django_filters.rest_framework.DjangoFilterBackend'],
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.LimitOffsetPagination',
    'PAGE_SIZE': 100,
}

# -----------------------------------------
# TLS FORWARDED
# -----------------------------------------
SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

# -----------------------------------------
# DJANGO TABLES2 TEMPLATE
# -----------------------------------------
DJANGO_TABLES2_TEMPLATE = "generics/table/table.html"


# -----------------------------------------
# django-dbbackup https://django-dbbackup.readthedocs.io/en/master/index.html
# -----------------------------------------
def backup_filename(databasename, servername, datetime, extension, content_type):
    return f"{content_type}-{datetime}.{extension}"


DBBACKUP_FILENAME_TEMPLATE = backup_filename
DBBACKUP_MEDIA_FILENAME_TEMPLATE = backup_filename
DBBACKUP_CLEANUP_KEEP = int(os.environ.get('DBBACKUP_CLEANUP_KEEP', 5))
DBBACKUP_CLEANUP_KEEP_MEDIA = int(os.environ.get('DBBACKUP_CLEANUP_KEEP', 5))
DBBACKUP_STORAGE = 'django.core.files.storage.FileSystemStorage'
DBBACKUP_STORAGE_OPTIONS = {'location': 'backup'}

# mysqldump_major_version = get_mysql_dump_major_version()
# print(f"mysqldump major version: {mysqldump_major_version}")
# if mysqldump_major_version < 10:
#     DBBACKUP_CONNECTORS = {
#         'default': {
#             'DUMP_SUFFIX': '--no-tablespaces --column-statistics=0',
#         }
#     }

if BACKUP_ENABLED:
    CELERY_BEAT_SCHEDULE["perform_backup"] = {
        "task": "service_catalog.tasks.perform_backup",
        "schedule": crontab(**get_celery_crontab_parameters_from_crontab_line(BACKUP_CRONTAB)),
    }

# -----------------------------------------
# SQUEST VERSION
# -----------------------------------------
try:
    repo = git.Repo(search_parent_directories=True)
    SQUEST_COMMIT_SHA6 = str(repo.head.object.hexsha)[0:6]
except git.exc.InvalidGitRepositoryError:
    SQUEST_COMMIT_SHA6 = "local dev"
print(f"SQUEST_VERSION: {__version__} - {SQUEST_COMMIT_SHA6}")

# -----------------------------------------
# Metrics
# -----------------------------------------
METRICS_ENABLED = str_to_bool(os.environ.get('METRICS_ENABLED', False))
METRICS_PASSWORD_PROTECTED = str_to_bool(os.environ.get('METRICS_PASSWORD_PROTECTED', True))
METRICS_AUTHORIZATION_USERNAME = os.environ.get('METRICS_AUTHORIZATION_USERNAME', 'admin')
METRICS_AUTHORIZATION_PASSWORD = os.environ.get('METRICS_AUTHORIZATION_PASSWORD', 'admin')

# -----------------------------------------
# Testing settings
# -----------------------------------------
if TESTING:
    import logging

    logging.disable(logging.CRITICAL)
    DEBUG = False
    TEMPLATE_DEBUG = False
    PASSWORD_HASHERS = [
        'django.contrib.auth.hashers.MD5PasswordHasher',
    ]
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': ':memory:',
        }
    }

CACHES = {
    "default": {
        "BACKEND": "django.core.cache.backends.redis.RedisCache",
        "LOCATION": f"redis://{REDIS_CACHE_USER}:{REDIS_CACHE_PASSWORD}@{REDIS_CACHE_HOST}:{REDIS_CACHE_PORT}/0",
    }
}
if TESTING or DEBUG or IS_MAKING_MIGRATION or IS_MIGRATING or IS_COLLECTING_STATIC or IS_INSERTING_DEFAULT_DATA:
    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.dummy.DummyCache',
        }
    }
SWAGGER_SETTINGS = {
    'SECURITY_DEFINITIONS': {
        'Basic': {
            'type': 'basic'
        },
        'Bearer (format: Bearer <token>)': {
            'type': 'apiKey',
            'name': 'Authorization',
            'in': 'header',
            'bearerFormat': 'Bearer',
            'scheme': 'bearer'
        }
    }
}

# -----------------------------------------
# Plugins
# -----------------------------------------
FIELD_VALIDATOR_PATH = "plugins/field_validators"
SURVEY_VALIDATOR_PATH = "plugins/survey_validators"

# -----------------------------------------
# SQL Debug
# -----------------------------------------
if SQL_DEBUG:
    LOGGING["loggers"]["django.db.backends"] = {
        "level": "DEBUG"
    }

print('[Settings] loaded')
